js/src/components/tree-select-control/index.js

```diff
@@ -169,30 +194,104 @@ const TreeSelectControl = ( {
 			);
 		};
 
-		const filterOption = ( option ) => {
-			if ( hasChildren( option ) ) {
-				option.children = option.children.filter( filterOption );
-				return !! option.children.length;
-			}
+		const descriptors = {
+			allLeaves: {
+				/**
+				 * Returns all (filtered) leaf options under this option.
+				 *
+				 * @return {InnerOption[]} All (filtered) leaf options under this option.
+				 */
+				get() {
+					if ( ! this.hasChildren ) {
+						return [];
+					}
+					return this.children.flatMap( ( option ) => {
+						return option.hasChildren ? option.allLeaves : option;
+					} );
+				},
+			},
+			checked: {
+				/**
+				 * Returns whether this option is checked.
+				 * A child option is checked its value is selected.
+				 * A parent option is checked if all leaves are checked.
+				 *
+				 * @return {boolean} True if checked, false otherwise.
+				 */
+				get() {
+					if ( this.hasChildren ) {
+						return this.allLeaves.every( ( opt ) => opt.checked );
+					}
+					return current.selectedValues.includes( this.value );
+				},
+			},
+			partialChecked: {
+				/**
+				 * Returns whether this option is partially checked.
+				 * A child option always returns false.
+				 * A parent option is partially checked if at least one but not all leaves are checked.
+				 *
+				 * @return {boolean} True if partially checked, false otherwise.
+				 */
+				get() {
+					if ( ! this.hasChildren ) {
+						return false;
+					}
+					return (
+						! this.checked &&
+						this.allLeaves.some(
+							( opt ) => opt.checked || opt.partialChecked
+						)
+					);
+				},
+			},
+			expanded: {
+				/**
+				 * Returns whether this option is expanded.
+				 * A child option always returns false.
+				 *
+				 * @return {boolean} True if expanded, false otherwise.
+				 */
+				get() {
+					return (
+						isFiltered ||
+						this.value === ROOT_VALUE ||
+						current.expandedValues.includes( this.value )
+					);
+				},
+			},
+		};
 
-			const match = option.label.toLowerCase().indexOf( filter );
+		const reduceOptions = ( acc, { children = [], ...option } ) => {
+			option.hasChildren = children.length > 0;
 
-			if ( match >= 0 ) {
-				option.label = highlightOptionLabel( option.label, match );
-				return true;
-			}
-		};
+			if ( option.hasChildren ) {
+				option.children = children.reduce( reduceOptions, [] );
 
-		if ( cachedFilteredOptions ) {
-			setFilteredOptions( cachedFilteredOptions );
-			return;
-		}
+				if ( option.children.length ) {
+					Object.defineProperties( option, descriptors );
+					acc.push( option );
+				}
+			} else {
+				if ( isFiltered ) {
+					const match = option.label.toLowerCase().indexOf( filter );
+					if ( match === -1 ) {
+						return acc;
+					}
+					option.label = highlightOptionLabel( option.label, match );
+				}
+
+				Object.defineProperties( option, descriptors );
+				acc.push( option );
+			}
```

Comment:
💅 

Just an idea, I guess the reducer could be simplified with: 

```
const descriptors = {
            hasChildren: {
				get() {
					return this.children?.length > 0;
				},
			},
			allLeaves: {
```
```
const reduceOptions = ( acc, { children = [], ...option } ) => {
			if ( children.length ) {
				option.children = children.reduce( reduceOptions, [] );
			} else if ( isFiltered ) {
				const match = option.label.toLowerCase().indexOf( filter );
				if ( match === -1 ) {
					return acc;
				}
				option.label = highlightOptionLabel( option.label, match );
			}

			if ( option.children?.length || ! children.length ) {
				Object.defineProperties( option, descriptors );
				acc.push( option );
			}
			return acc;
		};
```

Reply:
Thanks for the nice suggestion! Based on your example, I adjusted it to a similar implementation in https://github.com/woocommerce/google-listings-and-ads/pull/1522/commits/da745d8b36ba4e9e020ae00604f85d8fcf238630.

---

> WDYT of adding a small comment in the reducer?

The processing of this reducer is also included in the code comments above, so I guess it's already sufficient?

https://github.com/woocommerce/google-listings-and-ads/blob/afa8de0035bd66217caa96b5212fb51ff1ea2be9/js/src/components/tree-select-control/index.js#L163-L169